### PR TITLE
Project: Enumerables Method

### PR DESCRIPTION
**Pull request by taiwocoker**
## In this milestone:
I have learned how to rebuild enumerable methods which are:
-my_each
-my_each_with_index
-my_select
-my_all
-my_any
-my_none
-my_count
-my_map
-my_inject

### CODE REVIEW

**Review by rOluochKe — CHANGES_REQUESTED** [review]
Hello @taiwocoker 
### You have done great so far :clap: However, you have a few issues you need to solve before your project could be approved.
- [ ] Write a more descriptive readme file detailing the application, its installation and how to run it
- [ ] my_all? method is not passing all the test cases, it displays an error refer to the image attached
my_all
- Sample test cases for my_all? method `%w[ant bear cat].my_all? { |word| word.length >= 3 }  #=> true`  `%w[ant bear cat].my_all?(/t/)   #=> false`  `[1, 2i, 3.14].my_all?(Numeric)   #=> true`
- [ ] my_any? method raises errors when I run test cases, see image attached 
my_any
- Sample test cases ` [nil, true, 99].my_any?(Integer)   #=> true` , `[nil, true, 99].my_any?  #=> true` ,  `[nil, true, 99].any?(Integer)   #=> true`
- [ ] my_none also raises issues when test cases are run
my_none

### For test cases on enumerable and how to write them, check out this [documentation](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-all-3F)

### Make these changes and resubmit for another code review. Good luck.
- Happy coding!

By Raymond
slack @iamraymond

File: enumerables.rb [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+      end
+      return true
+    end
+    return true unless block_given?
+
+    my_each do |num|
+      return false unless yield(num)
+    end
+    true
+  end
```
Comment (maelfosso) [inline enumerables.rb]:
```rb
2.6.5 :010 > false_array = [1, false, 'hi', []] 
 => [1, false, "hi", []] 
2.6.5 :011 > false_array.all?
 => false 
2.6.5 :012 > false_array.my_all?
 => true 
2.6.5 :013 > 
```

File: enumerables.rb [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+      end
+      return false
+    end
+    return true unless block_given?
+
+    my_each do |num|
+      return true if yield(num)
+    end
+    false
+  end
```
Comment (maelfosso) [inline enumerables.rb]:
```rb
2.6.5 :015 > false_array = [nil, false, nil, false] 
 => [nil, false, nil, false] 
2.6.5 :016 > false_array.any?
 => false 
2.6.5 :017 > false_array.my_any?
 => true 
2.6.5 :018 > 
```

File: enumerables.rb [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+      end
+      return true
+    end
+    return false unless block_given?
+
+    my_each do |num|
+      return false if yield(num)
+    end
+    true
+  end
```
Comment (maelfosso) [inline enumerables.rb]:
```rb
2.6.5 :020 > false_array
 => [nil, false, nil, false] 
2.6.5 :021 > false_array.none?
 => true 
2.6.5 :022 > false_array.my_none?
 => false 
2.6.5 :023 >
```

File: enumerables.rb [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+    my_select { |x| yield x }.length
+  end
+
+  def my_map(my_proc = nil)
+    array = []
+    my_each { |x| array << my_proc.call(x) } if my_proc
+    my_each { |x| array << yield(x) } if block_given?
+
+    array
+  end
```
Comment (maelfosso) [inline enumerables.rb]:
```rb
2.6.5 :025 > array
 => [2, 6, 3, 8, 4, 1, 6, 5, 0, 0, 3, 7, 3, 7, 3, 2, 5, 5, 4, 1, 3, 6, 5, 1, 7, 3, 1, 2, 8, 2, 5, 8, 2, 1, 1, 8, 6, 0, 5, 7, 8, 3, 7, 3, 1, 6, 5, 3, 4, 2, 5, 8, 1, 2, 8, 7, 4, 1, 2, 1, 2, 6, 1, 4, 2, 2, 5, 6, 6, 7, 5, 5, 1, 6, 2, 4, 8, 7, 0, 4, 7, 4, 6, 0, 6, 1, 2, 5, 3, 5, 5, 0, 2, 2, 5, 3, 2, 3, 5, 1] 
2.6.5 :026 > array.count(0)
 => 6 
2.6.5 :027 > array.my_count(0)
 => 0 
2.6.5 :028 >
```

Comment (maelfosso) [inline enumerables.rb]:
```rb
2.6.5 :030 > array
 => [2, 6, 3, 8, 4, 1, 6, 5, 0, 0, 3, 7, 3, 7, 3, 2, 5, 5, 4, 1, 3, 6, 5, 1, 7, 3, 1, 2, 8, 2, 5, 8, 2, 1, 1, 8, 6, 0, 5, 7, 8, 3, 7, 3, 1, 6, 5, 3, 4, 2, 5, 8, 1, 2, 8, 7, 4, 1, 2, 1, 2, 6, 1, 4, 2, 2, 5, 6, 6, 7, 5, 5, 1, 6, 2, 4, 8, 7, 0, 4, 7, 4, 6, 0, 6, 1, 2, 5, 3, 5, 5, 0, 2, 2, 5, 3, 2, 3, 5, 1] 
2.6.5 :031 > array.map
 => #<Enumerator: [2, 6, 3, 8, 4, 1, 6, 5, 0, 0, 3, 7, 3, 7, 3, 2, 5, 5, 4, 1, 3, 6, 5, 1, 7, 3, 1, 2, 8, 2, 5, 8, 2, 1, 1, 8, 6, 0, 5, 7, 8, 3, 7, 3, 1, 6, 5, 3, 4, 2, 5, 8, 1, 2, 8, 7, 4, 1, 2, 1, 2, 6, 1, 4, 2, 2, 5, 6, 6, 7, 5, 5, 1, 6, 2, 4, 8, 7, 0, 4, 7, 4, 6, 0, 6, 1, 2, 5, 3, 5, 5, 0, 2, 2, 5, 3, 2, 3, 5, 1]:map> 
2.6.5 :032 > array.my_map
 => [] 
```

Should return an **Enumerator**

File: enumerables.rb [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+      ignore_first = true
+      accumulator = first
+    end
+    index = 0
+    each do |element|
+      accumulator = block.call(accumulator, element) unless ignore_first && index.zero?
+      index += 1
+    end
+    accumulator
+  end
```
Comment (maelfosso) [inline enumerables.rb]:
```rb
2.6.5 :035 > array
 => [2, 6, 3, 8, 4, 1, 6, 5, 0, 0, 3, 7, 3, 7, 3, 2, 5, 5, 4, 1, 3, 6, 5, 1, 7, 3, 1, 2, 8, 2, 5, 8, 2, 1, 1, 8, 6, 0, 5, 7, 8, 3, 7, 3, 1, 6, 5, 3, 4, 2, 5, 8, 1, 2, 8, 7, 4, 1, 2, 1, 2, 6, 1, 4, 2, 2, 5, 6, 6, 7, 5, 5, 1, 6, 2, 4, 8, 7, 0, 4, 7, 4, 6, 0, 6, 1, 2, 5, 3, 5, 5, 0, 2, 2, 5, 3, 2, 3, 5, 1] 
2.6.5 :036 > array.inject(:+)
 => 387 
2.6.5 :037 > array.my_inject(:+)
Traceback (most recent call last):
        7: from /home/maelfosso/.rvm/rubies/ruby-2.6.5/bin/irb:23:in `<main>'
        6: from /home/maelfosso/.rvm/rubies/ruby-2.6.5/bin/irb:23:in `load'
        5: from /home/maelfosso/.rvm/rubies/ruby-2.6.5/lib/ruby/gems/2.6.0/gems/irb-1.0.0/exe/irb:11:in `<top (required)>'
        4: from (irb):37
        3: from enumerables.rb:118:in `my_inject'
        2: from enumerables.rb:118:in `each'
        1: from enumerables.rb:119:in `block in my_inject'
NoMethodError (undefined method `call' for nil:NilClass)
2.6.5 :038 > 
```

**Review by maelfosso — CHANGES_REQUESTED** [review]
# Changes required :red_circle: 

Hello,
Thanks for the improvement done so far.
However, to move forward there are still bugs to fix. 

Kindly read the comments left below, fix the bugs and submit a new code review request
Thanks

Mael FOSSO

**Comment by lekegitrepo** [comment]
### Hi @taiwocoker :clap:
#### Well done :+1:  for implementing some of the required changes. Based on the last review you still need to fix the following before this project can be approved.

- my_any?
- my_none?
- my_map?
- my_inject?

**_kindly fix these and submit for review._**

### Reviewed by Temitayo Adeleke.

**Review by lekegitrepo — CHANGES_REQUESTED** [review]
#### Status: Request changes

File: enumerables.rb [hunk 6]
```diff
@@ -0,0 +1,135 @@
+# rubocop:disable Metrics/ModuleLength
+# rubocop:disable Metrics/PerceivedComplexity,Metrics/CyclomaticComplexity
+module Enumerable
+  def my_each
+    return to_enum(:my_each) unless block_given?
+
+    each do |num|
+      yield(num)
+    end
+  end
```
Comment (shshamim63) [inline enumerables.rb:10]: Based on the instructions of this project you should not use any enumerable methods inside your defined methods. But I can see you are using it in your my_each. Isn't it possible to complete my-each without using each inside it? I organized the code in the following form.
```ruby
def my_each
    return to_enum(:my_each) unless block_given?

    for item in self do
      yield(item)
    end
    self
end
```

**Review by shshamim63 — APPROVED** [review]
## Status :100: Approved
___
Hi @taiwocoker :wave:
You have covered all the changes asked by the previous reviewer. And all your methods are functioning properly. I just found a simple issue and I am marking it as an optional change.
___
### :clubs: Optional Changes :clubs: 
___
1. You should not use any enumerable methods inside your functions. But I can see the use of each inside my_each. Check the comment below.
___
##### This :+1: is from me. You can go ahead and merge it.
___
## :herb: Happy Coding :herb: 
### Shakhawat Hossain

**Pull request merged**